File: 0121-买卖股票的最佳时机/Solution2.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }
        // dp[i] 代表截止到i天所得最大利润
        // dp[i] = max(dp[i-1], prices[i]-minprice)
        int minprice = prices[0];
        int[] dp = new int[prices.length];
        dp[0] = 0;
        for(int i=1; i<prices.length; i++){
            // 决定是否在第i天卖出来计算截止到该天的最大利润
            dp[i] = Math.max(dp[i-1], prices[i]-minprice);
            // 更新minprice
            minprice = Math.min(minprice, prices[i]);
        }

        return dp[prices.length-1];
    }
}

File: 0084-柱状图中最大的矩形/solution2.py
from typing import *

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        area = 0
        low = min(heights)
        stack_forward = []
        stack_back = [0] * len(heights)
        for i in range(len(heights) - 1, -1, -1):
            j = i + 1
            while j < len(stack_back) and heights[i] < heights[j]:
                j = stack_back[j]
            if j < len(stack_back) and heights[i] == heights[j]:
                j = stack_back[j]
            stack_back[i] = j
        for i in range(len(heights)):
            forward = -1
            while stack_forward and heights[i] <= heights[stack_forward[-1]]:
                stack_forward.pop()
            if stack_forward:
                forward = stack_forward[-1]
            back = stack_back[i]
            area = max(area, heights[i] * (back - forward - 1))
            stack_forward.append(i)
        return area



    """
    def largestRectangleArea(self, heights: List[int]) -> int:
        area = 0
        low = min(heights)
        stack_forward = []
        # stack_back = []
        for i in range(len(heights)):
            forward, back = -1, len(heights)
            while stack_forward and heights[i] <= heights[stack_forward[-1]]: stack_forward.pop()
            if stack_forward:
                forward = stack_forward[-1]
            if heights[i] != low:
                back = i
                while back < len(heights) and heights[back] >= heights[i]: back += 1
            # while stack_back and heights[i] <= heights[stack_back[-1]]: stack_back.pop()
            # if stack_back:
            #     back = stack_back[-1]
            area = max(area, heights[i] * (back - forward - 1))
            stack_forward.append(i)
        return area
    """

    """
    # TLE
    def largestRectangleArea(self, heights: List[int]) -> int:
        area = 0
        for i in range(len(heights)):
            left, right = i, i
            while left >= 0 and heights[left] >= heights[i]: left -= 1
            while right < len(heights) and heights[i] <= heights[right]: right += 1
            area = max(area, heights[i] * (right - left - 1))
        return area
    """


a = [4, 2, 0, 3, 2, 4, 3, 4]
ans = Solution().largestRectangleArea(a)
print(ans)